### PR TITLE
secp256k1: configure with more feature flags

### LINKED ISSUE #18637: Enable additional flags for secp256k1
There are a few more flags we should enable on `secp256k1`:
- `--enable-experimental`
- `--enable-module-schnorr`
- `--enable-module-ecdh`

(I had to add these flags for the sake of something that depends on secp256k1 - I don't remember for sure, but I think it's bitcoinj.)

### PR DESCRIPTION
**Pull request by chris-martin**
###### Motivation for this change

Closes #18637
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/tools/security/secp256k1/default.nix [hunk 1]
```diff
@@ -16,7 +16,12 @@ stdenv.mkDerivation rec {
 
   buildInputs = [ autoconf automake libtool ];
 
-  configureFlags = [ "--enable-module-recovery" ];
+  configureFlags = [
+    "--enable-experimental"
```
Comment (rasendubi) [inline pkgs/tools/security/secp256k1/default.nix]: Does it make sense to make these flags optional? I mean, when I see `experimental`, I doubt it should be enabled by default.

File: pkgs/tools/security/secp256k1/default.nix [hunk 2]
```diff
@@ -16,7 +16,12 @@ stdenv.mkDerivation rec {
...
+    "--enable-module-ecdh"
+    "--enable-module-recovery"
+    "--enable-module-schnorr"
+  ];
 
   preConfigure = "./autogen.sh";
```
Comment (fpletz) [inline pkgs/tools/security/secp256k1/default.nix]: You should instead use `autoreconfHook` in `nativeBuildInputs`.

**Comment by rasendubi** [comment]
@chris-martin, are you still interested in the PR?

**Comment by chris-martin** [comment]
Oh yeah, sorry, I'll update this in a bit.

**Comment by rasendubi** [comment]
@chris-martin ping

**Comment by rasendubi** [comment]
@chris-martin, another friendly reminder

**Comment by chris-martin** [comment]
Sorry, I promise this is still on my to-do list!

**Comment by rasendubi** [comment]
@chris-martin no worries – I understand you ;)

**Comment by chris-martin** [comment]
Okay, updated!

* Bumped to latest version
* All flags are optional
* Added JNI flag
* Removed schorr flag (which apparently no longer exists)
* Using `autoreconfHook` as suggested

**Review by rasendubi — APPROVED** [review]

File: pkgs/tools/security/secp256k1/default.nix [hunk 3]
```diff
@@ -1,32 +1,51 @@
...
+
+# Enable ECDH shared secret computation (disabled by default because it is
+# experimental)
+, enableECDH ? false
+
+# Enable libsecp256k1_jni (disabled by default because it requires a jdk,
+# which is a large dependency)
+, enableJNI ? false
+
+, ... }:
```
Comment (fpletz) [inline pkgs/tools/security/secp256k1/default.nix]: Small nitpick: `...` is not necessary here as `callPackage` will do the magic for you.

File: pkgs/tools/security/secp256k1/default.nix [hunk 4]
```diff
@@ -1,32 +1,51 @@
...
+
+  nativeBuildInputs = [ autoreconfHook ];
 
-  configureFlags = [ "--enable-module-recovery" ];
+  configureFlags =
+    optionals enableECDH [ "--enable-module-ecdh" "--enable-experimental" ] ++
+    optionals enableRecovery [ "--enable-module-recovery" ] ++
+    optionals enableJNI [ "--enable-jni" ];
 
   preConfigure = "./autogen.sh";
```
Comment (fpletz) [inline pkgs/tools/security/secp256k1/default.nix]: You can remove this hook because `autoreconfHook` already does this in a new build phase.

**Review by fpletz — APPROVED** [review]

**Comment by chris-martin** [comment]
Updated:

* Removed `, ...`
* Removed `preConfigure = "./autogen.sh";`

**Comment by chris-martin** [comment]
I think this should be ready to go now.

**Pull request merged**